README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# igoR <a href='https://dieghernan.github.io/igoR/'><img src="man/figures/logo.png" align="right" height="139"/></a>

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/igoR)](https://CRAN.R-project.org/package=igoR)
[![CRAN
results](https://cranchecks.info/badges/worst/igoR)](https://cran.r-project.org/web/checks/check_results_igoR.html)
[![CRAN-Downloads](http://cranlogs.r-pkg.org/badges/grand-total/igoR)](https://cran.r-project.org/package=igoR)
[![r-universe](https://dieghernan.r-universe.dev/badges/igoR)](https://dieghernan.r-universe.dev/)
[![Rbuild-status](https://github.com/dieghernan/igoR/workflows/R-CMD-check/badge.svg)](https://github.com/dieghernan/igoR/actions)
[![codecov](https://codecov.io/gh/dieghernan/igoR/branch/main/graph/badge.svg?token=UH3VLTTTRE)](https://app.codecov.io/gh/dieghernan/igoR)
[![CodeFactor](https://www.codefactor.io/repository/github/dieghernan/igor/badge)](https://www.codefactor.io/repository/github/dieghernan/igor)
[![DOI](https://img.shields.io/badge/DOI-10.5281/zenodo.4442958-blue)](https://doi.org/10.5281/zenodo.4442958)
[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
[![status](https://tinyverse.netlify.com/badge/igoR)](https://CRAN.R-project.org/package=igoR)

<!-- badges: end -->

The goal of `igoR` is to provide access and to extract information from
the Intergovernmental Organizations Database (IGOs), version 3, provided
by the Correlates of War Project <https://correlatesofwar.org/>
(Pevehouse et al. (2020)\>).

The dataset (V3) includes information of **534** IGO from 1816 to 2014,
as well as membership information.

**Source**: [Intergovernmental Organizations
(v3)](https://correlatesofwar.org/data-sets/IGOs).

Documentation and vignettes on <https://dieghernan.github.io/igoR//>

Additionally, a distribution of the State System Membership (Correlates
of War Project (2017)) is included on this package.

## Installation

Install `igoR` from [**CRAN**](https://CRAN.R-project.org/package=igoR):

``` r
install.packages("igoR")
```

You can install the developing version from GitHub:

``` r
library(remotes)
install_github("dieghernan/igoR")
```

Alternatively, you can install `igoR` using the
[r-universe](https://dieghernan.r-universe.dev/ui#builds):

``` r
# Enable this universe
options(repos = c(
  dieghernan = "https://dieghernan.r-universe.dev",
  CRAN = "https://cloud.r-project.org"
))


install.packages("igoR")
```

## Basic usage

### Search an IGO by name:

Search all IGOs related with sugar:

``` r
library(igoR)

result_sugar <- igo_search("Sugar")
```

| ionum | ioname  | orgname                                    | longorgname                                                     | label                                                           | sdate | deaddate | dead | integrated | replaced | igocode | version | accuracyofpre1965membershipdates                      | sourcesandnotes | imputed | political | social | economic |
| ----: | :------ | :----------------------------------------- | :-------------------------------------------------------------- | :-------------------------------------------------------------- | ----: | -------: | ---: | ---------: | -------: | ------: | ------: | :---------------------------------------------------- | :-------------- | ------: | --------: | -----: | -------: |
|    40 | AMSC    | African/Malgasy Sugar Council              | African and Malagasy Sugar Council                              | African and Malagasy Sugar Council                              |  1966 |     1977 |    1 |          0 |        0 |      NA |     2.1 | Not applicable - created 1965 or later                |                 |       0 |         0 |      0 |        1 |
|  1920 | GLACSEC | Group of L/A & Carib. Sugar Exp. Countries | Group of Latin American and Caribbean Sugar Exporting Countries | Group of Latin American and Caribbean Sugar Exporting Countries |  1974 |     2001 |    1 |          0 |        0 |      NA |     2.3 | Not applicable - created 1965 or later                |                 |       0 |         1 |      0 |        0 |
|  3130 | ISuC    | Intl Sugar Council                         | International Sugar Council                                     | International Sugar Council                                     |  1937 |     1967 |    1 |          0 |        0 |      91 |     3.0 | Within 5 years                                        |                 |       0 |         0 |      1 |        0 |
|  4290 | SugU    | Sugar Union                                | Sugar Union                                                     | Sugar Union                                                     |  1902 |     1913 |    1 |          0 |        0 |      29 |     3.0 | Completely accurate, except a few minor uncertainties |                 |       0 |         0 |      0 |        1 |

### IGO members:

Composition of the [European Economic
Community](https://en.wikipedia.org/wiki/European_Economic_Community)
over time:

``` r

eec_code <- igo_search("EEC", exact = TRUE)

# Founding
eec_init <- igo_members(eec_code$ioname, year = eec_code$sdate)
```

| ioname | ccode | state       | statenme                | year | value | category        | orgname                     |
| :----- | ----: | :---------- | :---------------------- | ---: | ----: | :-------------- | :-------------------------- |
| EEC    |   211 | belgium     | Belgium                 | 1958 |     1 | Full Membership | European Economic Community |
| EEC    |   220 | france      | France                  | 1958 |     1 | Full Membership | European Economic Community |
| EEC    |   325 | italy       | Italy                   | 1958 |     1 | Full Membership | European Economic Community |
| EEC    |   212 | luxembourg  | Luxembourg              | 1958 |     1 | Full Membership | European Economic Community |
| EEC    |   210 | netherlands | Netherlands             | 1958 |     1 | Full Membership | European Economic Community |
| EEC    |   260 | wgermany    | German Federal Republic | 1958 |     1 | Full Membership | European Economic Community |

EEC, members (1958)

``` r
# Latest date
eec_end <- igo_members(eec_code$ioname)
```

| ioname | ccode | state       | statenme       | year | value | category        | orgname                     |
| :----- | ----: | :---------- | :------------- | ---: | ----: | :-------------- | :-------------------------- |
| EEC    |   211 | belgium     | Belgium        | 1992 |     1 | Full Membership | European Economic Community |
| EEC    |   390 | denmark     | Denmark        | 1992 |     1 | Full Membership | European Economic Community |
| EEC    |   220 | france      | France         | 1992 |     1 | Full Membership | European Economic Community |
| EEC    |   255 | germany     | Germany        | 1992 |     1 | Full Membership | European Economic Community |
| EEC    |   350 | greece      | Greece         | 1992 |     1 | Full Membership | European Economic Community |
| EEC    |   205 | ireland     | Ireland        | 1992 |     1 | Full Membership | European Economic Community |
| EEC    |   325 | italy       | Italy          | 1992 |     1 | Full Membership | European Economic Community |
| EEC    |   212 | luxembourg  | Luxembourg     | 1992 |     1 | Full Membership | European Economic Community |
| EEC    |   210 | netherlands | Netherlands    | 1992 |     1 | Full Membership | European Economic Community |
| EEC    |   235 | portugal    | Portugal       | 1992 |     1 | Full Membership | European Economic Community |
| EEC    |   230 | spain       | Spain          | 1992 |     1 | Full Membership | European Economic Community |
| EEC    |   200 | uk          | United Kingdom | 1992 |     1 | Full Membership | European Economic Community |

EEC, members (1992)

## Recommended packages

  - `countrycode` for converting country names and codes across
    different systems (ISO3, Eurostat, World Bank, UN, FIPS/GEC, etc..)
  - `dplyr` for data manipulation.

## Citation

To cite the ‘igoR’ package in publications use this entry:

Hernangómez, D (2021). igoR: Intergovernmental Organizations Database. R
package version 0.1.3. <https://doi.org/10.5281/zenodo.4442958>

You can also use:

``` r
print(citation("igoR"), bibtex = TRUE)
#> 
#> To cite the 'igoR' package in publications use this entry:
#> 
#>   Hernangómez, D (2021). igoR: Intergovernmental Organizations
#>   Database. R package version 0.1.3.
#>   https://doi.org/10.5281/zenodo.4442958
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Misc{,
#>     title = {igoR: Intergovernmental Organizations Database},
#>     author = {Diego Hernangómez},
#>     year = {2021},
#>     note = {R package version 0.1.3},
#>     url = {https://dieghernan.github.io/igoR/},
#>     doi = {10.5281/zenodo.4442958},
#>   }
#> 
#> Please cite also:
#> 
#>   Pevehouse JC, Nordstrom T, McManus RW, Jamison AS. Tracking
#>   organizations in the world: The Correlates of War IGO Version 3.0
#>   datasets. Journal of Peace Research. 2020;57(3):492-503.
#>   doi:10.1177/0022343319881175
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Article{,
#>     author = {Jon CW Pevehouse and Timothy Nordstrom and Roseanne W McManus and Anne Spencer Jamison},
#>     title = {Tracking organizations in the world: The Correlates of War IGO Version 3.0 datasets},
#>     journal = {Journal of Peace Research},
#>     volume = {57},
#>     number = {3},
#>     pages = {492-503},
#>     year = {2020},
#>     doi = {10.1177/0022343319881175},
#>     url = {https://doi.org/10.1177/0022343319881175},
#>     eprint = {https://doi.org/10.1177/0022343319881175},
#>     abstract = {This article summarizes the Correlates of War Intergovernmental Organizations (IGO) Version 3.0 datasets. The new datasets include information about the population of IGOs in the international system and state participation in those formal international institutions from 1816 to 2014. Consistent with Versions 2.0 and 2.3, Version 3.0 of the IGO data comes in three forms: country-year, IGO-year, and joint dyadic membership.},
#>   }
```

## References

<div id="refs" class="references">

<div id="ref-cow_2017">

Correlates of War Project. 2017. “State System Membership List, v2016.”
<https://correlatesofwar.org/>.

</div>

<div id="ref-doi:10.1177/0022343319881175">

Pevehouse, Jon CW, Timothy Nordstrom, Roseanne W McManus, and Anne
Spencer Jamison. 2020. “Tracking Organizations in the World: The
Correlates of War IGO Version 3.0 Datasets.” *Journal of Peace Research*
57 (3): 492–503. <https://doi.org/10.1177/0022343319881175>.

</div>

</div>


NEWS.md
# igoR 0.1.3

-   Update docs.

# igoR 0.1.2

-   "Mapping IGOs" article compiled to vignette.
-   Update docs. Now in markdown format using `roxygen2md::roxygen2md()`.
-   Moved tests to `testthat`.

# igoR 0.1.1

-   Remove lifecycle badge.

# igoR 0.1.0

-   Vectorize `igo_members()`
-   Generalize search ignoring case.
-   Add `igo_state_membership()`
-   `cow_country_codes` now is internal.
-   Added `igoR::states2016` in replacement of `cow_country_codes`.
-   Add `igo_dyadic()`
-   Add `igo_search_states()`
-   More vignettes on the website

# igoR 0.0.0.9000

-   Pre-release


R/igo_dyadic.R
#' @title Extract the Joint Membership of a pair of Countries across IGOs.
#' @name igo_dyadic
#' @description Dyadic version of the data. The unit of observation is a dyad
#' of countries. It provides a summary of the joint memberships of two IGOs
#' over time.
#'
#' @return A coded data frame representing the years and country dyad (rows)
#' and the IGOs selected (columns). See Details
#' @seealso [state_year_format3],  [states2016],
#' [igo_search()].
#' @source [**Codebook Version 3
#' IGO Data**](https://correlatesofwar.org/data-sets/IGOs) for full reference.
#' @references
#' Pevehouse, Jon CW, Timothy Nordstrom, Roseanne W McManus, and Anne Spencer
#' Jamison. "Tracking Organizations in the World: The Correlates of War
#' IGO Version 3.0 Datasets." *Journal of Peace Research* 57, no. 3
#' (May 2020): 492-503. \doi{10.1177/0022343319881175}.
#'
#' @param country1 A single state, used as a base of comparison. It could be
#' any valid name or code of a state as specified on [states2016].
#' @param country2 A state of vector of states to be compared with
#' `country1`.
#' @param year Year to be assessed, an integer or an array of year.
#' @param ioname Optional. `ioname` or vector of `ioname`
#' corresponding to the IGOs to be assessed. If `NULL` (the default),
#' all IGOs would be extracted. See codes on [igo_search()].
#' @details
#' This function tries to replicate the information contained in the original
#' file distributed by The Correlates of War Project
#' (`dyadic_format3.dta`). That file is not included in this package due
#' to its size.
#'
#' The result is a data frame containing the common years of the states
#' selected via `country1, country2, year` by rows.
#'
#' An additional column `dyadid`, computed as `(1000*ccode1)+ccode2`
#' is provided in order to identify relationships.
#'
#' For each IGO selected via `ioname` (or all if the default
#' option has been used) a column (using lowercase `ioname` as
#' identifier) is provided with the following code system:
#' \tabular{cc}{
#'   **Category** \tab **Numerical Value**\cr
#'     No Joint Membership \tab 0 \cr
#'     Joint Full Membership \tab 1 \cr
#'     Missing data \tab -9 \cr
#'     State Not System Member \tab -1 \cr
#'     }
#' If one state in an IGO is a full member but the other is an associate
#' member or observer, that IGO is not coded as a joint membership.
#'
#'  **Differences with the original dataset**
#'
#'  There are some differences on the results provided by this function and the
#'  original dataset on some IGOs regarding the "Missing Data" (-9) and
#'  "State Not System Member" (-1). However it is not clear how to fully
#'  replicate those values.
#'
#' See
#' [**Codebook Version 3 IGO Data**](https://correlatesofwar.org/data-sets/IGOs)
#'
#'
#' @export
#' @examples
#' USA_CAN <- igo_dyadic("USA", "Spain")
#' nrow(USA_CAN)
#' ncol(USA_CAN)
#'
#' # Using custom parameters
#' igo_dyadic(
#'   country1 = "France",
#'   country2 =
#'     c("Sweden", "Austria"),
#'   year = 1992:1995,
#'   ioname = "EU"
#' )
igo_dyadic <- function(country1,
                       country2,
                       year = 1816:2014,
                       ioname = NULL) {
  # Check countries
  if (length(country1) != 1) {
    stop("country1 should be a single value")
  }

  country1 <- igoR::igo_search_states(country1)$stateabb
  country2 <- unique(igoR::igo_search_states(country2)$stateabb)

  # Initial clean up
  country2 <- country2[country2 != country1]

  if (length(country2) == 0) {
    stop(
      "Codes selected correspond to the same country: ",
      country1
    )
  }

  if (length(country2) == 1) {
    # Get IGO info

    all_igos <- igoR::state_year_format3

    min <- min(all_igos$year)
    max <- max(all_igos$year)

    # Filter by years

    all_igos <- all_igos[all_igos$year %in% year, ]

    # Check years

    if (nrow(all_igos) == 0) {
      stop(
        "The value(s) of year are not valid. Years range: ",
        min,
        " - ",
        max
      )
    }


    # Check ionames if selected

    if (!is.null(ioname)) {
      coligos <- colnames(all_igos)

      ioname <- unique(tolower(ioname))
      colsel <- match(ioname, coligos)

      if (anyNA(colsel)) {
        warning(
          "ioname(s) ",
          paste0("'",
            ioname[is.na(colsel)], "'",
            collapse = ","
          ),
          " not valid"
        )
      }

      if (all(is.na(colsel))) {
        stop("Execution halted. No valid ionames used.")
      }

      colsel <- c(1:2, colsel[!is.na(colsel)])
      all_igos <- all_igos[, colsel]
    }

    # Extract countries

    c1 <- igoR::igo_search_states(country1)
    c2 <- igoR::igo_search_states(country2)

    # nocov start
    if (c1$ccode == c2$ccode) {
      stop(
        "Codes selected correspond to the same country: ",
        c1$statenme
      )
    }
    # nocov end

    # Extract igos by country
    c1_igos <- all_igos[all_igos$ccode == c1$ccode, ]
    colnames(c1_igos) <- gsub("ccode", "ccode1", colnames(c1_igos))
    colnames(c1_igos) <- gsub("state", "state1", colnames(c1_igos))

    c2_igos <- all_igos[all_igos$ccode == c2$ccode, ]
    colnames(c2_igos) <- gsub("ccode", "ccode2", colnames(c2_igos))
    colnames(c2_igos) <- gsub("state", "state2", colnames(c2_igos))

    # Master table with common years
    master <- merge(c1_igos[, 1:2], c2_igos[, 1:2])
    master$dyadid <- master$ccode1 * 1000 + master$ccode2
    # Check
    if (nrow(master) == 0) {
      stop(
        "No common records on the years selected. ",
        "One country (or both) does not exist on that range.",
        "Countries selected: ",
        c1$statenme,
        ", ",
        c2$statenme
      )
    }

    # Refilter country tables
    c1_igos <- c1_igos[c1_igos$year %in% master$year, ]

    c2_igos <- c2_igos[c2_igos$year %in% master$year, ]


    # Recode table
    igos_colnames <- colnames(c1_igos)
    igos_colnames <-
      igos_colnames[!igos_colnames %in% c("ccode1", "year", "state1")]

    c1_matrix <- as.matrix(c1_igos[, igos_colnames])
    c2_matrix <- as.matrix(c2_igos[, igos_colnames])
    end_matrix <-
      matrix(
        data = NA,
        nrow = nrow(c1_matrix),
        ncol = ncol(c1_matrix)
      )
    colnames(end_matrix) <- igos_colnames

    dims <- dim(end_matrix)

    # Create coding
    for (r in seq_len(dims[1])) {
      for (c in seq_len(dims[2])) {
        vals <- c(c1_matrix[r, c], c2_matrix[r, c])

        if (any(vals == -9)) {
          finalvalue <- -9
        } else if (any(vals == -1)) {
          finalvalue <- -1
        } else if (all(vals == 1)) {
          finalvalue <- 1
        } else if (any(vals %in% c(0, 2, 3))) {
          finalvalue <- 0
        } else {
          # Safe coding
          # nocov start
          finalvalue <- -9
          # nocov end
        }

        end_matrix[r, c] <- finalvalue
      }
    }

    # Handle NAs
    end_matrix[is.na(end_matrix)] <- -9

    end_igos <- as.data.frame(end_matrix)
    end_igos$year <- master$year

    # Regenerate table
    colnames(c1) <- paste0(colnames(c1), 1)
    colnames(c2) <- paste0(colnames(c2), 2)

    master <- merge(master, c1)
    master <- merge(master, c2)
    master <- merge(master, end_igos)

    # Reorder rows and cols

    colorder <-
      unique(c("dyadid", colnames(c1), colnames(c2), colnames(master)))
    master <- master[, colorder]
    master <- master[order(master$year), ]

    rownames(master) <- NULL

    return(master)
  } else {
    # Vectorized part for country2----
    df <-
      igo_dyadic(
        country1 = country1,
        country2 = country2[1],
        year = year,
        ioname = ioname
      )

    dflen <- seq_len(length(country2))[-1]

    for (i in dflen) {
      df <-
        rbind(
          df,
          igo_dyadic(
            country1 = country1,
            country2 = country2[i],
            year = year,
            ioname = ioname
          )
        )
    }

    return(df)
  }
}
